src/libopensc/reader-cryptotokenkit.m

```diff
@@ -0,0 +1,426 @@
+	  		if ([priv->tksmartcardslot.ATR.bytes length] > SC_MAX_ATR_SIZE)
+	   			return SC_ERROR_INTERNAL;
+	  		reader->atr.len = [priv->tksmartcardslot.ATR.bytes length];
+	  		memcpy(reader->atr.value, (unsigned char*) [priv->tksmartcardslot.ATR.bytes bytes], reader->atr.len);
+	  		break;
+	 	case TKSmartCardSlotStateMissing:
+	  		// Slot is no longer known to the system.
+	  		reader->flags &= ~(SC_READER_CARD_PRESENT);
+	  		reader->flags |= SC_READER_REMOVED;
+	  		r = SC_ERROR_READER_DETACHED;
+	  		break;
+	 	case TKSmartCardSlotStateEmpty:
+	  		/// The slot is empty, no card is inserted.
+	  		reader->flags &= ~SC_READER_CARD_PRESENT;
+	  		break;
+	 	default:
+	  		r = SC_ERROR_UNKNOWN;
+	  		break;
+	}
+
+	if ((old_flags & SC_READER_CARD_PRESENT) == (reader->flags & SC_READER_CARD_PRESENT))
+	 	reader->flags &= ~SC_READER_CARD_CHANGED;
+	else
+	 	reader->flags |= SC_READER_CARD_CHANGED;
+
+	sc_log(reader->ctx, "card %s%s",
+			reader->flags & SC_READER_CARD_PRESENT ? "present" : "absent",
+		   reader->flags & SC_READER_CARD_CHANGED ? ", changed": "");
+
+	if (r == SC_SUCCESS)
+	 	r = reader->flags;
+
+	LOG_FUNC_RETURN(reader->ctx, r);
+}
+
+static int ctk_proto_to_opensc(TKSmartCardProtocol proto)
+{
+	switch (proto) {
+	   	case TKSmartCardProtocolT0:
+			return SC_PROTO_T0;
+	   	case TKSmartCardProtocolT1:
+			/* fall through */
+	   	case TKSmartCardProtocolT15:
+			return SC_PROTO_T1;
+	   	case TKSmartCardProtocolAny:
+			return SC_PROTO_ANY;
+	   	default:
+			return 0;
+	}
+}
+
+static void ctk_set_proto(sc_reader_t *reader)
+{
+ 	struct cryptotokenkit_private_data *priv = reader->drv_data;
+ 	if (priv->tksmartcard) {
+  		reader->active_protocol = ctk_proto_to_opensc(priv->tksmartcard.currentProtocol);
+  		if (priv->tksmartcard.allowedProtocols & TKSmartCardProtocolAny) {
+   			reader->supported_protocols = ctk_proto_to_opensc(TKSmartCardProtocolAny);
+  		} else {
+   			if (priv->tksmartcard.allowedProtocols & TKSmartCardProtocolT0)
+				reader->supported_protocols |= ctk_proto_to_opensc(TKSmartCardProtocolT0);
+   			if (priv->tksmartcard.allowedProtocols & TKSmartCardProtocolT1)
+				reader->supported_protocols |= ctk_proto_to_opensc(TKSmartCardProtocolT1);
+   			if (priv->tksmartcard.allowedProtocols & TKSmartCardProtocolT15)
+				reader->supported_protocols |= ctk_proto_to_opensc(TKSmartCardProtocolT1);
+  		}
+ 	}
+}
+
+static int cryptotokenkit_connect(sc_reader_t *reader)
+{
+	struct cryptotokenkit_private_data *priv = reader->drv_data;
+
+	if (!priv->tksmartcard) {
+		priv->tksmartcard = [priv->tksmartcardslot makeSmartCard];
+	}
+
+	if (!priv->tksmartcard || ![priv->tksmartcard valid])
+		return SC_ERROR_CARD_NOT_PRESENT;
+
+	/* attempt to detect protocol in use T0/T1/RAW */
+	ctk_set_proto(reader);
+
+	return SC_SUCCESS;
+}
+
+static int cryptotokenkit_disconnect(sc_reader_t * reader)
+{
+	struct cryptotokenkit_private_data *priv = reader->drv_data;
+
+	priv->tksmartcard = NULL;
```

Comment:
Well, I already spent way too long to figure out this kind of detail for some language I don't really want to use. If you make a suggestion, I'll try it out, otherwise I'll leave the code with what I've tested.